### ISSUE TITLE
Failing test in akka.http.impl.engine.ws.Utf8CodingSpecs after replacing triple equals with shouldEqual assertion.

### ISSUE THREAD

**Issue by nikom1337** [akka/akka-http#1760]
`akka.http.impl.engine.ws.Utf8CodingSpecs` test class is the last one which is misusing ===. After using shouldEqual we're getting randomly 2 different failures:

```
[ERROR] [01/08/2018 13:16:21.922] [Utf8CodingSpecs-akka.actor.default-dispatcher-4] [akka://Utf8CodingSpecs/user/StreamSupervisor-0/flow-12-0-headSink] Error in stage [Utf8Encoder]: Unexpected UTF-16 surrogate continuation
java.lang.IllegalArgumentException: Unexpected UTF-16 surrogate continuation
	at akka.http.impl.engine.ws.Utf8Encoder$$anon$1.step$1(Utf8Encoder.scala:53)
	at akka.http.impl.engine.ws.Utf8Encoder$$anon$1.onPush(Utf8Encoder.scala:72)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:512)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:404)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:584)
	at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:554)
	at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:675)
	at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:723)
	at akka.actor.Actor$class.aroundPreStart(Actor.scala:528)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:666)
	at akka.actor.ActorCell.create(ActorCell.scala:591)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:462)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:484)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.run(Mailbox.scala:223)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[info] Utf8CodingSpecs:
[info] Utf8 decoding/encoding
[info] - work for all codepoints *** FAILED ***
[info]   RuntimeException was thrown during property evaluation.
[info]     Message: Trying to await result of failed Future, see the cause for the original problem.
[info]     Occurred when passed generated values (
[info]       arg0 = -8514 // 3 shrinks
[info]     )
```

or

```
[info] Utf8CodingSpecs:
[info] Utf8 decoding/encoding
[info] - work for all codepoints *** FAILED ***
[info]   TestFailedException was thrown during property evaluation.
[info]     Message: "[?]?" did not equal "[?]?"
[info]     Location: (Utf8CodingSpecs.scala:39)
[info]     Occurred when passed generated values (
[info]       arg0 = 930497
[info]     )
[error] java.nio.charset.MalformedInputException: Input length = 1
[error] 	at java.nio.charset.CoderResult.throwException(CoderResult.java:281)
[error] 	at sun.nio.cs.StreamEncoder.implWrite(StreamEncoder.java:285)
[error] 	at sun.nio.cs.StreamEncoder.write(StreamEncoder.java:125)
[error] 	at sun.nio.cs.StreamEncoder.write(StreamEncoder.java:135)
[error] 	at java.io.Writer.write(Writer.java:157)
[error] 	at scala.xml.XML$.write(XML.scala:112)
[error] 	at scala.xml.XML$.$anonfun$save$2(XML.scala:94)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:224)
[error] 	at scala.xml.XML$.save(XML.scala:94)
[error] 	at sbt.JUnitXmlTestsListener.writeSuite(JUnitXmlTestsListener.scala:197)
[error] 	at sbt.JUnitXmlTestsListener.endGroup(JUnitXmlTestsListener.scala:186)
[error] 	at sbt.TestRunner.$anonfun$run$7(TestFramework.scala:118)
[error] 	at sbt.TestRunner.$anonfun$run$7$adapted(TestFramework.scala:118)
[error] 	at sbt.TestFramework$.$anonfun$safeForeach$1(TestFramework.scala:143)
[error] 	at sbt.TestFramework$.$anonfun$safeForeach$1$adapted(TestFramework.scala:142)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:929)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:929)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
[error] 	at sbt.TestFramework$.safeForeach(TestFramework.scala:142)
[error] 	at sbt.TestRunner.safeListenersCall(TestFramework.scala:128)
[error] 	at sbt.TestRunner.run(TestFramework.scala:118)
[error] 	at sbt.TestFramework$$anon$2$$anonfun$$lessinit$greater$1.$anonfun$apply$1(TestFramework.scala:262)
[error] 	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:233)
[error] 	at sbt.TestFramework$$anon$2$$anonfun$$lessinit$greater$1.apply(TestFramework.scala:262)
[error] 	at sbt.TestFramework$$anon$2$$anonfun$$lessinit$greater$1.apply(TestFramework.scala:262)
[error] 	at sbt.TestFunction.apply(TestFramework.scala:271)
[error] 	at sbt.Tests$.$anonfun$toTask$1(Tests.scala:281)
[error] 	at sbt.std.Transform$$anon$3.$anonfun$apply$2(System.scala:46)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:66)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:271)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] java.nio.charset.MalformedInputException: Input length = 1
```

After this issue is fixed, #1136 can be also closed.

Refs: #1136

**Comment by jrudolph** [comment]
Wow, great catch, @nikom1337. There was indeed a bug in the utf16 surrogate pair decoder of the UTF8 encoder.

Fix in #1761.

**Comment by jrudolph** [comment]
(Luckily the unicode range above 0x20000 is not yet widely used, so probably no one stumbled over the actual effects so far).
